pretyping/unification.ml

```diff
@@ -1031,19 +1043,23 @@ let rec unify_0_with_initial_metas (sigma,ms,es as subst : subst0) conv_at_top e
       | None -> (* some undefined Metas in cN *) None
       | Some n1 ->
          (* No subterm restriction there, too much incompatibilities *)
-         let sigma =
+         let uprob =
            if opt.with_types then
              try (* Ensure we call conversion on terms of the same type *)
                let tyM = get_type_of curenv ~lax:true sigma m1 in
                let tyN = get_type_of curenv ~lax:true sigma n1 in
-               check_compatibility curenv CUMUL flags substn tyM tyN
+               check_compatibility_ustate curenv CUMUL flags substn tyM tyN
              with RetypeError _ ->
-               (* Renounce, maybe metas/evars prevents typing *) sigma
-           else sigma
+               (* Renounce, maybe metas/evars prevents typing *) UnivProblem.Set.empty
+           else UnivProblem.Set.empty
          in
-        match infer_conv ~pb ~ts:convflags curenv sigma m1 n1 with
-        | Some sigma ->
-          Some (sigma, metasubst, evarsubst)
+        match infer_conv_ustate ~pb ~ts:convflags curenv sigma m1 n1 with
+        | Some uprob' ->
+          let uprob = UnivProblem.Set.(force (union uprob uprob')) in
+          begin match Evd.add_universe_constraints sigma uprob with
+          | sigma -> Some (sigma, metasubst, evarsubst)
+          | exception (Univ.UniverseInconsistency _) -> None
```

Comment:
forcing here seems to fix the hott error, but I don't understand why catching UniversesDiffer -> None produces an error at Defined.